lib/puppet/parser/compiler.rb

```diff
@@ -190,6 +207,44 @@ def evaluate_node_classes
     evaluate_classes(classes_without_params, @node_scope || topscope)
   end
 
+  def evaluate_applications
+    @applications.each do |app|
+      mapping = app.parameters[:nodes].value
+      components = []
+      mapping.each do |k,v|
+        k = [k] unless k.is_a?(Array)
+        v = [v] unless v.is_a?(Array)
+        k.each do |key|
+          v.each do |val|
+            key, val = val, key if val.type == "Node"
+            if key.type == "Node" && key.title == node.name
+              components << val
+            end
+          end
+        end
+      end
+      # @todo lutter 2015-01-28: We use @current_app and @curremt_comp to
+      # signal to the evaluator that we are in the middle of evaluating an
+      # application.
+      begin
+        @current_app = app
+        @current_components = components
```

Comment:
I prefer these state instance variables (`@current_app` and `@current_components`) to be defined and documented instead of introduced here in the middle of things and just having a comment above explaining what they are. They could be introduced as private attr_reader.